crates/typos-cli/src/bin/typos-cli/report.rs

```diff
@@ -280,6 +287,132 @@ impl Report for PrintJson {
+                    .physical_location(sarif::PhysicalLocationBuilder::default()
+                        .artifact_location(
+                            sarif::ArtifactLocationBuilder::default()
+                                .uri(msg.context.as_ref().map_or("", |context| match context {
+                                    Context::File(context) => context.path.to_str().unwrap_or(""),
+                                    _ => "",
+                                }))
+                                .build().unwrap()
+                        )
+                        .region(
+                            sarif::RegionBuilder::default()
+                                .start_line(line_num as i64)
+                                .start_column(column_number as i64)
+                                .end_column((column_number + msg.typo.len()) as i64)
+                                .build().unwrap()
+                        )
+                        .build().unwrap())
+                    .build().unwrap();
+
+
+                let result = sarif::ResultBuilder::default()
+                    .level(ResultLevel::Error.to_string())
+                    .message(MessageBuilder::default()
+                        .markdown(message)
+                        .build().unwrap())
+                    .locations(vec![location])
+                    .build().unwrap();
+
+                self.results.lock().unwrap().borrow_mut().push(result);
+            }
+            Message::Parse(msg) => {
+                self.error.lock().unwrap().replace(Some(msg.data.to_owned()));
+            }
+            Message::Error(msg) => {
+                self.error.lock().unwrap().replace(Some(msg.msg.clone()));
+            }
+            _ => { unreachable!() }
+        }
+
+        Ok(())
+    }
+
+    fn finalize(&self) -> Result<(), std::io::Error> {
+        let mut sarif_builder = sarif::SarifBuilder::default();
+        sarif_builder.version(sarif::Version::V2_1_0.to_string())
+            .schema(sarif::SCHEMA_URL);
+
+
+        let tool = sarif::ToolBuilder::default()
+            .driver(sarif::ToolComponentBuilder::default()
+                .name("typos")
+                .information_uri("https://github.com/crate-ci/typos")
+                .build().unwrap())
+            .build().unwrap();
+
+        let mut run_builder = sarif::RunBuilder::default();
+        run_builder.tool(tool)
+            .results(self.results.lock().unwrap().borrow().clone());
+
+
+        if let Some(error) = &self.error.lock().unwrap().borrow().as_deref() {
+            run_builder.invocations(vec![
+                sarif::InvocationBuilder::default()
+                    .process_start_failure_message(error.to_string())
+                    .build().unwrap()
+            ]);
+        }
+
+        let sarif = sarif_builder.build().unwrap();
+
+        write!(stdout().lock(), "{}", serde_json::to_string(&sarif).unwrap())?;
```

Comment:
`to_string` or `to_string_pretty`?  Which does sarif expect?

Reply:
sarif basically a json file. I think both is ok as it only requires machine-readable

Reply:
If SARIF uses json and not jsonlines, let's go with `to_string_pretty` to make this easier for humans

Reply:
updated

Reply:
Ok